### PR TITLE
Add midpoint elements polygon

### PR DESCRIPTION
**Pull request by swvickers**
Adding midpoint elements to polygon line segments.
These midpoint elements will serve as visual cues to convert straight line segments to curved lines.

MidpointElements

### CODE REVIEW

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts [hunk 1]
```diff
@@ -145,6 +145,39 @@ export class RegionData implements IRegionData, IMovable, IResizable {
...
+            // closing line segment from last to first point
+            segments.push([points[pointsLength - 1], points[0]])
+        }
+        return segments;
+    }
+
+    public getLineMidpoints(): Point2D[] {
+        const lines = this.getLineSegments();
+        return lines.map(line => {
+            const x = line[0].x - ((line[0].x - line[1].x) / 2);
```
Comment (Man-MSFT) [inline src/canvastools/ts/CanvasTools/Core/RegionData.ts:175]: line[i].x is always greater than line[i+1].x ?

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts [hunk 2]
```diff
@@ -145,6 +145,39 @@ export class RegionData implements IRegionData, IMovable, IResizable {
...
+        }
+        if (this.regionType == RegionDataType.Polygon) {
+            // closing line segment from last to first point
+            segments.push([points[pointsLength - 1], points[0]])
+        }
+        return segments;
+    }
+
+    public getLineMidpoints(): Point2D[] {
+        const lines = this.getLineSegments();
```
Comment (Man-MSFT) [inline src/canvastools/ts/CanvasTools/Core/RegionData.ts:173]: I may be misusing GitHub's code review, but I only see this function (getLineSegments) used in one place.
1. Could we make it private?
2. We basically do two iterations here, first iteration to create the segments, second iteration to get midpoints. But it can be made into one iteration, maybe combine the two fxns?

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts [hunk 3]
```diff
@@ -145,6 +145,39 @@ export class RegionData implements IRegionData, IMovable, IResizable {
...
+            const nextPointIdx = i + 1;
+            if (nextPointIdx < pointsLength) {
+                segments.push([points[i], points[nextPointIdx]])
+            }
+        }
+        if (this.regionType == RegionDataType.Polygon) {
+            // closing line segment from last to first point
+            segments.push([points[pointsLength - 1], points[0]])
+        }
+        return segments;
```
Comment (Man-MSFT) [inline src/canvastools/ts/CanvasTools/Core/RegionData.ts:169]: If we isolate this code, it seems independent of state (if we just pass in `points`). Wonder if it makes sense to have a geometryutils or lineUtils class for pure mathematical calcs? Then we can pass `return lineSegments(this.regionPoints, this.regionType)`

File: src/canvastools/ts/CanvasTools/Interface/IEventDescriptor.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@
  * @remarks Used as an array type to specify a collection of listeners that should
  * added to each of the elements in another array.
  */
-export interface IEventDescriptor {
+ export interface IEventDescriptor<T extends PointerEvent | MouseEvent | KeyboardEvent | WheelEvent> {
```
Comment (Man-MSFT) [inline src/canvastools/ts/CanvasTools/Interface/IEventDescriptor.ts:6]: Do we have tslint/eslint rules that would have caught this?

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts [hunk 5]
```diff
@@ -145,6 +145,39 @@ export class RegionData implements IRegionData, IMovable, IResizable {
         return area;
     }
 
+    public getLineSegments(): Array<[Point2D, Point2D]> {
```
Comment (Man-MSFT) [inline src/canvastools/ts/CanvasTools/Core/RegionData.ts:148]: Tests coming in next push?

File: src/canvastools/ts/CanvasTools/Region/Component/MidpointComponent.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        event: "click",
+        base: midpoint.node,
+        listener: (e: MouseEvent) => {
+          e.stopPropagation();
+          console.log("Midpoint component clicked", e);
+        },
+        bypass: false,
+      },
+    ];
+    this.subscribeToEvents(listeners);
```
Comment (Man-MSFT) [inline src/canvastools/ts/CanvasTools/Region/Component/MidpointComponent.ts:139]: If we end up calling the function twice then we subscribe twice? Should/can we make idempotent?

File: webpack.config.js [hunk 7]
```diff
@@ -42,15 +42,11 @@ var webpackSettings = {
...
-    settings = webpackSettings['dev'];
-}
-
 module.exports = function (env) {
-    var settings = webpackSettings[env.mode];
+    const mode = (env && env.mode) || "prod";
+    var settings = webpackSettings[mode];
     if (settings == undefined) {
-        settings = webpackSettings['dev'];
+        settings = webpackSettings['prod'];
```
Comment (Man-MSFT) [inline webpack.config.js:49]: Are these changes intended to check in, or just things leftover from dev/test?

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts (hunk 1 above)
Comment (Man-MSFT) [reply src/canvastools/ts/CanvasTools/Core/RegionData.ts:175]: If not, we need to sort. If yes, who maintains order? Is it just inserted in order and our code just always maintains it?

**Review by Man-MSFT — CHANGES_REQUESTED** [review]
Left comments about code structure, and some general q's for my own knowledge.

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts (hunk 1 above)
Comment (swvickers) [reply src/canvastools/ts/CanvasTools/Core/RegionData.ts:175]: Points are added into the array in the order they are added during the polygon creation.
Lines are calculated by order of the points in the array, creating a "DAG" of sorts with a single path of vertices and edges.

This method starts at the beginning of this path and measure out a halfway point between each vertex along the way.

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts (hunk 3 above)
Comment (swvickers) [reply src/canvastools/ts/CanvasTools/Core/RegionData.ts:169]: Yea, I think there are a lot of opportunities to extract common code into utilities. I've been hewing more closely to the existing code architecture (highly object oriented with 3+ layers of classes/abstracts interfaces) which has quite a few very similar class/code functions with slight tweaks between them.
I'm hoping to have some time at the end to refactor to DRY it up.

File: src/canvastools/ts/CanvasTools/Interface/IEventDescriptor.ts (hunk 4 above)
Comment (swvickers) [reply src/canvastools/ts/CanvasTools/Interface/IEventDescriptor.ts:6]: There are rules, but no enforcement. TSlint was warning me about it, so I fixed it. I'm not really focusing on fixing and refactoring primarily, so I'm keeping them as opportunistic updates at the moment. Trying to avoid changing too many contracts/interfaces which would cascade into lots of code changes spread around.

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts (hunk 5 above)
Comment (swvickers) [reply src/canvastools/ts/CanvasTools/Core/RegionData.ts:148]: I'll look into the test situation, there don't seem to be any in the application right now so the test framework may need to be setup or fixed up.

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts (hunk 3 above)
Comment (Man-MSFT) [reply src/canvastools/ts/CanvasTools/Core/RegionData.ts:169]: Cool sounds good. Not a blocking comment

File: src/canvastools/ts/CanvasTools/Region/Component/MidpointComponent.ts (hunk 6 above)
Comment (swvickers) [reply src/canvastools/ts/CanvasTools/Region/Component/MidpointComponent.ts:139]: I think these subscribe calls are only called in constructors, so any new nodes have listeners attached to them. I'll double check.

File: src/canvastools/ts/CanvasTools/Interface/IEventDescriptor.ts (hunk 4 above)
Comment (Man-MSFT) [reply src/canvastools/ts/CanvasTools/Interface/IEventDescriptor.ts:6]: Got it. Definitely fine to defer it from this PR, but we may want to turn on stricter lint rules maybe warning as error, in future.

File: webpack.config.js (hunk 7 above)
Comment (swvickers) [reply webpack.config.js:49]: Planning on checking them in. Default build settings should be prod unless otherwise specified as a general good practice. I think the build, test and dev tasks need to be fixed up a bit.

File: src/canvastools/ts/CanvasTools/Core/RegionData.ts (hunk 5 above)
Comment (Man-MSFT) [reply src/canvastools/ts/CanvasTools/Core/RegionData.ts:148]: Yes, the lint werror thing, and this lack of tests especially, I think we should definitely create some tasks to pick up later for fixing.

File: webpack.config.js (hunk 7 above)
Comment (Man-MSFT) [reply webpack.config.js:49]: Feel free to jot these EE comments down as ADO items!
